chain/chain/src/thread_pool.rs

```diff
@@ -0,0 +1,107 @@
+use parking_lot::Mutex;
+use std::collections::VecDeque;
+use std::sync::Arc;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::time::Duration;
+use thiserror::Error;
+use thread_priority::{RealtimeThreadSchedulePolicy, ThreadBuilder, ThreadSchedulePolicy};
+
+pub(crate) trait ThreadLimit {
+    fn max_threads(&self) -> usize;
+}
+
+/// TODO: Implement `ThreadLimit` which returns the current number of tracked shards
+pub(crate) type ConstLimit = usize;
+
+impl ThreadLimit for ConstLimit {
+    fn max_threads(&self) -> usize {
+        *self
+    }
+}
+
+type Job = Box<dyn FnOnce() + Send + 'static>;
+type IdleQueue = Arc<Mutex<VecDeque<oneshot::Sender<Option<Job>>>>>;
+
+pub(crate) struct ThreadPool<C: ThreadLimit> {
+    name: &'static str,
+    limit: C,
+    idle_timeout: Duration,
+    num_threads: AtomicUsize,
+    idle_queue: IdleQueue,
+}
+
+#[derive(Error, Debug)]
+pub(crate) enum Error {
+    #[error("thread limit reached for pool {name}: {limit}")]
+    ThreadLimitReached { name: &'static str, limit: usize },
+    #[error("spawning thread failed: {0}")]
+    Spawn(#[from] std::io::Error),
+}
+
+impl<C: ThreadLimit> ThreadPool<C> {
+    pub(crate) fn new(name: &'static str, idle_timeout: Duration, limit: C) -> Self {
```

Comment:
It seems like an overkill to have the `ThreadLimit` trait and pass it here as such. Can you make it simpler? Can you just use a regular int parameter for the thread limit? 

In case you do have a good reason for it, can you improve naming a bit? Maybe `ThreadLimit: ThreadLimitTrait`?

Reply:
hmm actually I wonder if in case of forks we need 2x #shards

Reply:
I wouldn't mind not having a limit at all personally

Reply:
> I wouldn't mind not having a limit at all personally

We need some sort of safeguard if there's a bug in our code that spawns more tasks than necessary. Otherwise, we risk silently overloading the CPU with too many threads.

Reply:
> We need some sort of safeguard if there's a bug in our code that spawns more tasks than necessary. Otherwise, we risk silently overloading the CPU with too many threads.

I don't think the safeguard needs to be coded in, code review, testing and so forth should be sufficient.

Reply:
In that case I will remove the `ThreadLimit` trait for now and just use an `usize`.